src/cli/src/profile/sync/cloud/login_agent.rs

```diff
@@ -0,0 +1,255 @@
+            .set_header("Authorization", &*creds.token)
+            .await?;
+        match response.status() {
+            StatusCode::Ok => {
+                info!("Successfully authenticated with token!");
+                error!("TODO, implement profile download");
+                Ok(())
+            }
+            StatusCode::Unauthorized => {
+                warn!("Failed to download profile, token is expired or invalid");
+                Err(CloudError::Unauthorized)
+            }
+            StatusCode::NotFound => {
+                warn!("Failed to download a profile for this user");
+                Err(CloudError::ProfileNotFound)
+            }
+            _ => Err(CloudError::Unexpected),
+        }
+    }
+
+    /// Attempts to use a username and password to login to Fluvio Cloud.
+    ///
+    /// If this succeeds, the LoginAgent will save the Fluvio Cloud
+    /// credentials in a session to be used later.
+    pub async fn authenticate(
+        &mut self,
+        email: String,
+        password: String,
+    ) -> Result<(), CloudError> {
+        let login = LoginRequest {
+            email: email.clone(),
+            password,
+        };
+        let mut response = surf::post(format!("{}/api/v1/loginUser", &self.remote))
+            .body_json(&login)
+            .unwrap()
+            .await?;
+
+        match response.status() {
+            StatusCode::Ok => {
+                let creds = response.body_json::<Credentials>().await?;
+                self.save_credentials(creds).await?;
+                Ok(())
+            }
+            _ => {
+                warn!("Failed to login");
+                Err(CloudError::AuthenticationError(email))
+            }
+        }
+    }
+
+    async fn save_credentials(&mut self, creds: Credentials) -> Result<(), CloudError> {
+        // Save credentials in agent session
+        self.session.replace(creds);
+        // Save credentials to disk
+        self.session.as_ref().unwrap().try_save(&self.path).await?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Serialize)]
+struct LoginRequest {
+    email: String,
+    password: String,
+}
+
+#[derive(Debug, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(rename_all = "camelCase")]
+struct Credentials {
+    id: String,
+    token: String,
+}
+
+impl Credentials {
+    /// Try to load credentials from disk
+    async fn try_load<P: AsRef<Path>>(path: P) -> Result<Self, CloudError> {
+        let file_str = fs::read_to_string(path).await?;
+        let creds: Credentials =
+            toml::from_str(&*file_str).map_err(|_| CloudError::UnableToReadCredentials)?;
+        Ok(creds)
+    }
+
+    /// Try to save credentials to disk
+    async fn try_save<P: AsRef<Path>>(&self, path: P) -> Result<(), IoError> {
+        fs::create_dir_all(path.as_ref().parent().unwrap()).await?;
+        fs::write(path, toml::to_string(self).unwrap().as_bytes()).await
```

Comment:
same...

Reply:
So with this particular unwrap, I think it's safe to leave because we know for a fact that `Self: Serialize` and it should be impossible for a `Serialize` type to fail this. I added a comment to explain why this unwrap is safe.